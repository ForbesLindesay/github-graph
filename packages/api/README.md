# @github-graph/api

API client for GitHub's GraphQL API that automatically merges multiple requests into one where possible, saving you from hitting the API rate limits.

## Installation

```
yarn add @github-graph/api
```

## Usage

```ts
import Client, {auth, gql, getMethod} from '@github-graph/api';

const getStargazers = getMethod<
  {repository: {nameWithOwner: string; stargazers: {totalCount: number}}},
  {owner: string; name: string}
>(gql`
  query GetStargazers($owner: String!, $name: String!) {
    repository(owner: $owner, name: $name) {
      nameWithOwner
      stargazers {
        totalCount
      }
    }
  }
`);


const getWatchers = getMethod<
  {repository: {nameWithOwner: string; watchers: {totalCount: number}}},
  {owner: string; name: string}
>(gql`
  query GetWatchers($owner: String!, $name: String!) {
    repository(owner: $owner, name: $name) {
      nameWithOwner
      watchers {
        totalCount
      }
    }
  }
`);

const client = new Client({
  auth: auth.createTokenAuth(GITHUB_TOKEN),
});

await Promise.all([
  getStargazers(client, {owner: 'pugjs', name: 'pug'}),
  getWatchers(client, {owner: 'pugjs', name: 'pug'})),
  getStargazers(client, {owner: 'ForbesLindesay', name: 'atdatabases'}),
  getWatchers(client, {owner: 'ForbesLindesay', name: 'atdatabases'}),
]);
```

Unlike other libraries, that would send 4 separate queries to the backend. This library will only send one:

```ts
{
  query: `
    query ($owner: String!, $name: String!, $b: String!, $c: String!) {
      repository(owner: $owner, name: $name) {
        nameWithOwner
        stargazers {
          totalCount
        }
        watchers {
          totalCount
        }
      }
      b: repository(owner: $b, name: $c) {
        nameWithOwner
        stargazers {
          totalCount
        }
        watchers {
          totalCount
        }
      }
    }
  `,
  variables: {
    b: 'ForbesLindesay',
    c: 'atdatabases',
    name: 'pug',
    owner: 'pugjs',
  },
}
```

This is 4 times fewer tokens used from your rate limit.
