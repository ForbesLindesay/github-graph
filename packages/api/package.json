{
  "name": "@github-graph/api",
  "version": "0.0.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "repository": "git@github.com:ForbesLindesay/github-graph.git",
  "author": "Forbes Lindesay <forbes@lindesay.co.uk>",
  "license": "MIT",
  "scripts": {
    "build": "tsc"
  },
  "dependencies": {
    "@octokit/auth": "^2.0.0",
    "@octokit/request": "^5.3.2",
    "graphql": "^14.6.0",
    "graphql-merge-unmerge": "^2.0.0",
    "graphql-tag": "^2.10.3",
    "universal-user-agent": "^5.0.0"
  },
  "devDependencies": {
    "@forbeslindesay/tsconfig": "^2.0.0",
    "@types/node": "^13.9.1",
    "typescript": "^3.8.3"
  },
  "publishConfig": {
    "access": "public"
  }
}
