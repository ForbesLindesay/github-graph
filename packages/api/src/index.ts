// tslint:disable-next-line: no-implicit-dependencies
import {AuthInterface} from '@octokit/types';
import {request} from '@octokit/request';
import {getUserAgent} from 'universal-user-agent';
import {Batch} from 'graphql-merge-unmerge';
import {DocumentNode} from 'graphql/language/ast';
import {print} from 'graphql';
import gql from 'graphql-tag';
import * as auth from '@octokit/auth';
import {Octokit} from '@octokit/rest';
import takeToken, {
  BucketOptions,
  BucketState,
} from '@authentication/rate-limit/lib/bucket';

export {auth};
export {gql};

export interface RateLimitOptions extends BucketOptions {
  /**
   * Defaults to 30 seconds
   */
  maxDelay?: number;
}
export type Options = {
  maxBatchSize?: number;
  onRequest?: (request: {query: string; variables: any}) => void;
  onResponse?: (
    request: {query: string; variables: any},
    response: {data?: any; errors?: readonly any[]},
  ) => void;
  onBatchRequest?: (request: {query: string; variables: any}) => void;
  onBatchResponse?: (
    request: {query: string; variables: any},
    response: {
      headers: any;
      status: number;
      url: any;
      data: {data?: any; errors?: any[]};
    },
  ) => void;
} & (
  | {
      auth: Pick<AuthInterface<any, any>, 'hook'>;
      rateLimitOptions?: RateLimitOptions;
      userAgent?: string;
      request?: undefined;
    }
  | {
      auth?: undefined;
      userAgent?: undefined;
      request: (options: EndpointOptions) => Promise<OctoKitResponse<any>>;
    }
);
const VERSION = require('../package.json').version;
const USER_AGENT = `github-graph-api/${VERSION} ${getUserAgent()}`;

export type Method<TResult, TArgs> = {} extends TArgs
  ? (client: GitHubClient, args?: TArgs) => Promise<TResult>
  : (client: GitHubClient, args: TArgs) => Promise<TResult>;
export function getMethod<TResult, TArgs>(
  doc: DocumentNode,
): Method<TResult, TArgs> {
  return (async (client: GitHubClient, args?: TArgs): Promise<TResult> =>
    client.query(doc, args)) as Method<TResult, TArgs>;
}

export class GraphqlError extends Error {
  constructor(
    request: {query: string; variables: any},
    response: {data?: any; errors?: readonly any[]},
  ) {
    const message = response.errors![0].message;
    super(message);

    Object.assign(this, response, {request});
    this.name = 'GraphqlError';

    // Maintains proper stack trace (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }
}

// tslint:disable-next-line: no-implicit-dependencies
export type OctoKitResponse<T> = import('@octokit/types').OctokitResponse<T>;
// tslint:disable-next-line: no-implicit-dependencies
type PaginateInterface = import('@octokit/plugin-paginate-rest').PaginateInterface;
// tslint:disable-next-line: no-implicit-dependencies
type RestEndpointMethods = Octokit;
export type RestApi = Readonly<
  {paginate: PaginateInterface} & RestEndpointMethods
>;
// tslint:disable-next-line: no-implicit-dependencies
type EndpointOptions = import('@octokit/types').EndpointOptions;
export default class GitHubClient {
  private _batch: Batch | null = null;
  private _batchSize: number = 0;
  private readonly _options: Options;
  public readonly request: (
    options: EndpointOptions,
  ) => Promise<OctoKitResponse<any>>;
  public readonly rest: RestApi;
  constructor(options: Options) {
    if (options.auth) {
      this._options = options;
      const rateLimitOptions: RateLimitOptions | undefined =
        options.rateLimitOptions;
      let rateLimitState: BucketState | null = null;
      this.request = async (_options) => {
        if (rateLimitOptions) {
          const now = Date.now();
          const newRateLimitState = takeToken(rateLimitState, rateLimitOptions);
          if (
            newRateLimitState.timestamp - now >
            (rateLimitOptions.maxDelay || 30_000)
          ) {
            const err: any = new Error(
              `You have hit the rate limit you set when constructing the GitHubClient. You can make another request in ${Math.floor(
                (newRateLimitState.timestamp - now) / 1000,
              )} seconds`,
            );
            err.code = 'RATE_LIMIT_EXCEEDED';
            throw err;
          }
          rateLimitState = newRateLimitState;
          if (newRateLimitState.timestamp > now) {
            await new Promise((resolve) => {
              setTimeout(resolve, newRateLimitState.timestamp - now);
            });
          }
        }
        return await request({
          ..._options,
          headers: {
            'user-agent': options.userAgent || USER_AGENT,
            ..._options.headers,
          },
          request: {hook: options.auth.hook},
        });
      };
    } else {
      this._options = options;
      this.request = options.request;
    }
    this.rest = new Octokit({request: options.request});
  }
  private readonly _processQueue = async () => {
    if (this._batch) {
      const b = this._batch;
      this._batch = null;
      this._batchSize = 0;
      await b.run();
    }
  };

  public async query(query: DocumentNode, variables: any = {}): Promise<any> {
    if (this._batchSize >= (this._options.maxBatchSize || 100)) {
      void this._processQueue();
    }
    this._batchSize++;
    if (!this._batch) {
      this._batch = new Batch(async (q) => {
        let attempts = 0;
        while (true) {
          const req = {
            query: print(q.query),
            variables: q.variables,
          };
          if (this._options.onBatchRequest) this._options.onBatchRequest(req);
          const response = await this.request({
            ...req,
            method: 'POST',
            url: '/graphql',
          });

          if (this._options.onBatchResponse)
            this._options.onBatchResponse(req, response);
          if (
            response.data.errors?.length === 1 &&
            response.data.errors[0].type === 'RATE_LIMITED'
          ) {
            if (attempts++ > 3) {
              throw new GraphqlError(req, response);
            }
            await new Promise((resolve) => {
              setTimeout(resolve, attempts * 5000);
            });
            continue;
          }
          return response.data;
        }
      });
      void Promise.resolve(null).then(this._processQueue);
    }
    let request: {query: string; variables: any} | undefined;
    const getRequest = () =>
      request || (request = {query: print(query), variables});
    if (this._options.onRequest) {
      this._options.onRequest(getRequest());
    }
    const response = await this._batch?.queue({query, variables});
    if (this._options.onResponse) {
      this._options.onResponse(getRequest(), response);
    }
    if (response.errors && response.errors.length) {
      throw new GraphqlError(getRequest(), response);
    }
    return response.data;
  }
}
