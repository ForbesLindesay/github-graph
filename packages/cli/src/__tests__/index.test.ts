import generate from '../';
import {writeFileSync} from 'fs';

test('example', () => {
  const output = generate(`${__dirname}/example.graphql`);
  writeFileSync(
    __dirname + '/example.ts',
    generate(`${__dirname}/example.graphql`),
  );
  expect(output).toMatchInlineSnapshot(`
    "/* tslint:disable */
    /* eslint-disable */
    // @generated
    // This file was automatically generated and should not be edited.

    import {getMethod, gql} from '@github-graph/api';

    //==============================================================
    // START Enums and Input Objects
    //==============================================================

    /**
     * Autogenerated input type of AddStar
     */
    export interface AddStarInput {
      starrableId: string;
      clientMutationId?: string | null;
    }

    //==============================================================
    // END Enums and Input Objects
    //==============================================================


    // ====================================================
    // GraphQL query operation: GetLogin
    // ====================================================

    export interface GetLogin_viewer {
      /**
       * The username used to login.
       */
      login: string;
    }

    export interface GetLogin {
      /**
       * The currently authenticated user.
       */
      viewer: GetLogin_viewer;
    }

    export const getLogin = getMethod<GetLogin, {}>(gql\`
    query GetLogin {
      viewer {
        login
      }
    }
    \`);

    // ====================================================
    // GraphQL query operation: GetStargazers
    // ====================================================

    export interface GetStargazers_repository_stargazers {
      /**
       * Identifies the total count of items in the connection.
       */
      totalCount: number;
    }

    export interface GetStargazers_repository {
      /**
       * The repository's name with owner.
       */
      nameWithOwner: string;
      /**
       * A list of users who have starred this starrable.
       */
      stargazers: GetStargazers_repository_stargazers;
    }

    export interface GetStargazers {
      /**
       * Lookup a given repository by the owner and repository name.
       */
      repository: GetStargazers_repository | null;
    }

    export interface GetStargazersVariables {
      owner: string;
      name: string;
    }

    export const getStargazers = getMethod<GetStargazers, GetStargazersVariables>(gql\`
    query GetStargazers($owner: String!, $name: String!) {
      repository(owner: $owner, name: $name) {
        nameWithOwner
        stargazers {
          totalCount
        }
      }
    }
    \`);

    // ====================================================
    // GraphQL query operation: GetWatchers
    // ====================================================

    export interface GetWatchers_repository_watchers {
      /**
       * Identifies the total count of items in the connection.
       */
      totalCount: number;
    }

    export interface GetWatchers_repository {
      /**
       * The repository's name with owner.
       */
      nameWithOwner: string;
      /**
       * A list of users watching the repository.
       */
      watchers: GetWatchers_repository_watchers;
    }

    export interface GetWatchers {
      /**
       * Lookup a given repository by the owner and repository name.
       */
      repository: GetWatchers_repository | null;
    }

    export interface GetWatchersVariables {
      owner: string;
      name: string;
    }

    export const getWatchers = getMethod<GetWatchers, GetWatchersVariables>(gql\`
    query GetWatchers($owner: String!, $name: String!) {
      repository(owner: $owner, name: $name) {
        nameWithOwner
        watchers {
          totalCount
        }
      }
    }
    \`);

    // ====================================================
    // GraphQL mutation operation: AddStar
    // ====================================================

    export interface AddStar_addStar_starrable {
      id: string;
    }

    export interface AddStar_addStar {
      /**
       * The starrable.
       */
      starrable: AddStar_addStar_starrable | null;
    }

    export interface AddStar {
      /**
       * Adds a star to a Starrable.
       */
      addStar: AddStar_addStar | null;
    }

    export interface AddStarVariables {
      input: AddStarInput;
    }

    export const addStar = getMethod<AddStar, AddStarVariables>(gql\`
    mutation AddStar($input: AddStarInput!) {
      addStar(input: $input) {
        starrable {
          id
        }
      }
    }
    \`);

    // ====================================================
    // GraphQL query operation: GetWithFragment
    // ====================================================

    export interface GetWithFragment_repository_stargazers {
      /**
       * Identifies the total count of items in the connection.
       */
      totalCount: number;
    }

    export interface GetWithFragment_repository_watchers {
      /**
       * Identifies the total count of items in the connection.
       */
      totalCount: number;
    }

    export interface GetWithFragment_repository {
      /**
       * The repository's name with owner.
       */
      nameWithOwner: string;
      /**
       * A list of users who have starred this starrable.
       */
      stargazers: GetWithFragment_repository_stargazers;
      /**
       * A list of users watching the repository.
       */
      watchers: GetWithFragment_repository_watchers;
    }

    export interface GetWithFragment {
      /**
       * Lookup a given repository by the owner and repository name.
       */
      repository: GetWithFragment_repository | null;
    }

    export const getWithFragment = getMethod<GetWithFragment, {}>(gql\`
    query GetWithFragment {
      repository(owner: \\"pugjs\\", name: \\"pug\\") {
        ...repo
      }
    }
    fragment repo on Repository {
      nameWithOwner
      stargazers {
        totalCount
      }
      watchers {
        totalCount
      }
    }
    \`);

    // ====================================================
    // GraphQL fragment: repo
    // ====================================================

    export interface repo_stargazers {
      /**
       * Identifies the total count of items in the connection.
       */
      totalCount: number;
    }

    export interface repo_watchers {
      /**
       * Identifies the total count of items in the connection.
       */
      totalCount: number;
    }

    export interface repo {
      /**
       * The repository's name with owner.
       */
      nameWithOwner: string;
      /**
       * A list of users who have starred this starrable.
       */
      stargazers: repo_stargazers;
      /**
       * A list of users watching the repository.
       */
      watchers: repo_watchers;
    }
    "
  `);
});

test('example syntax error', () => {
  expect(() =>
    generate(`${__dirname}/example-syntax-error.graphql`),
  ).toThrowErrorMatchingSnapshot();
});

test('example schema error', () => {
  expect(() =>
    generate(`${__dirname}/example-schema-error.graphql`),
  ).toThrowErrorMatchingSnapshot();
});
