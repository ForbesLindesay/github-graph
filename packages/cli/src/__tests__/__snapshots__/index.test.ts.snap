// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example schema error 1`] = `
"Cannot query field \\"logn\\" on type \\"User\\"

  1 | query GetLogin {
  2 |   viewer {
> 3 |     logn
    |     ^
  4 |   }
  5 | }
  6 | 
"
`;

exports[`example syntax error 1`] = `
"Syntax Error: Expected Name, found <EOF>

  3 |     login
  4 |   
> 5 | }
    |  ^
"
`;
